### ISSUE TITLE
sam deploy CLI behavior is different depending on the use of the -t flag

### ISSUE THREAD

**Issue by dsandor** [aws/aws-sam-cli#3616]
### Description:
If you specify a `template.yaml` filename when invoking `sam deploy` the CLI packages up the source in the Lambda CodeUri directory instead of the build artifact. Removing the -t flag from the same command results in the correct behavior.

### Steps to reproduce:
Create a hello-world golang lambda using a Makefile based build:
1. Run `sam build -t template.other.yaml` 
2. Run `sam deploy --no-execute-changeset --stack-name delete-me --resolve-s3 -t template.other.yaml`
3. Inspect the contents of the zip file uploaded to S3.

When you inspect the zip file contents of the Lambda function archive uploaded to S3 by SAM CLI you will note that the source code was zipped up and uploaded.

Now do the same thing but do not use the `-t` command line option.
1. cp template.other.yaml template.yaml
1. Run `sam build` 
2. Run `sam deploy --no-execute-changeset --stack-name delete-me --resolve-s3`
3. Inspect the contents of the zip file uploaded to S3.

Now when you look at the zip file archive that was built you will notice that the built binary exists in the zip file instead of the source code.  This is how it is supposed to work.

TLDR; If you use the `-t` option with `sam deploy` it does not use the built artifacts from the `sam build` command. However, if you do not use the `-t` option everything works as intended.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: mac osx monteray
2. `sam --version`: SAM CLI, version 1.35.0
3. AWS region: us-east-1

**Comment by dsandor** [comment]
I did a further test on this and noted that the problem I am experiencing is explicitly occurring when the NAME of the template file is not `template.yaml`.  So I can use the `-t` option on `sam deploy` but if the name of the template is anything other than `template.yaml` the `sam build` artifacts are ignored.

**Comment by wchengru** [comment]
Thanks for raising this issue!

The `-t` option in `sam deploy` only deploy the template and the local resources that it points to. By default, the `sam deploy` command deploy with the template within `.aws-sam` directory in your project directory.

This is documented: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-deploy.html

To deploy the resource with specified template, please use `sam build --template` and then `sam deploy` instead of running `sam deploy --template.

**Comment by wchengru** [comment]
I will close this issue because this is an expected behaviour. Please reopen it if you have any further questions or requests.
